subsys/shell/modules/date_service.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2020 Nick Ward
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <shell/shell.h>
+#include <init.h>
+#include <string.h>
+#include <posix/time.h>
+#include <sys/timeutil.h>
+
+#define HELP_NONE "[none]"
+#define HELP_DATE_SET \
+	"<year: 1900-9999> <month: 1-12> <day: 1-31> " \
+	"<hour: 0-23> <minute: 0-59> <second: 0-60>"
+
+static void date_print(const struct shell *shell, struct tm *tm)
+{
+	shell_print(shell,
+		    "%d-%02d-%02d "
+		    "%02d:%02d:%02d UTC",
+		    tm->tm_year + 1900,
+		    tm->tm_mon + 1,
+		    tm->tm_mday,
+		    tm->tm_hour,
+		    tm->tm_min,
+		    tm->tm_sec);
+}
+
+static int cmd_date_set(const struct shell *shell, size_t argc, char **argv)
+{
+	struct tm tm = {0};
+	struct timespec tp;
+	int ret;
+
+	ARG_UNUSED(argc);
+
+	tm.tm_year = strtol(argv[1], NULL, 10);
+	if ((tm.tm_year < 1900) || (tm.tm_year > 9999)) {
+		shell_error(shell, "Invalid year");
+		return -EINVAL;
+	}
+	tm.tm_year -= 1900;
+
+	tm.tm_mon = strtol(argv[2], NULL, 10);
+	if ((tm.tm_mon < 1) || (tm.tm_mon > 12)) {
+		shell_error(shell, "Invalid month");
+		return -EINVAL;
+	}
+
+	tm.tm_mon -= 1;
+
+	tm.tm_mday = strtol(argv[3], NULL, 10);
+	/* Check day field against maximum month length */
+	if ((tm.tm_mday < 1) || (tm.tm_mday > 31)) {
+		shell_error(shell, "Invalid day");
+		return -EINVAL;
+	}
+	tm.tm_hour = strtol(argv[4], NULL, 10);
+	if ((tm.tm_hour < 0) || (tm.tm_hour > 23)) {
+		shell_error(shell, "Invalid hour");
+		return -EINVAL;
+	}
+	tm.tm_min = strtol(argv[5], NULL, 10);
+	if ((tm.tm_min < 0) || (tm.tm_min > 59)) {
+		shell_error(shell, "Invalid minute");
+		return -EINVAL;
+	}
+	tm.tm_sec = strtol(argv[6], NULL, 10);
+	/* Note range allows for a leap second */
```

Comment:
Thank you for respecting leap seconds even though nobody's ever going to enter one by hand.